### DISCUSSION TITLE
border-opacity doesn't work

### DISCUSSION THREAD

**Discussion by Florian-crg** [tailwindlabs/tailwindcss#5757] [Help] [answered]
**What version of Tailwind CSS are you using?**
    "tailwindcss": "^2.2.4"

**What build tool (or framework if it abstracts the build tool) are you using?**
NextJS  11.0.0

**What version of Node.js are you using?**

v12.22.1

**What browser are you using?**

Chrome / Firefox 

**What operating system are you using?**
M1

**Reproduction URL**

https://i.imgur.com/NHgGTsq.png

**Describe your issue**

The border opacity isn't working. It is a normal project using tailwind with nextJS. 
JIT mod is enabled.

**Comment by adamwathan** [comment, marked as answer]
Hey! It definitely works in general so must be something specific to your setup:

https://play.tailwindcss.com/mb4EaD7UON?file=css

Please show the actual code and there's a better chance of being able to help 👍🏻

**Reply by Florian-crg** [reply]
You can see in the photo linked that the code the border opacity is set and  that it doesn't work. What do you need ? My tailwind config file ? Or something else ? 

Here is the code : 
**ProductCard.module.css**
```
.header {
  @apply relative bottom-0  border-opacity-0 border-t border-primary-2 rounded-b-xl h-28 flex items-center md:justify-center text-black;
}
```
**ProductCard.tsx**
```
...
import s from './ProductCard.module.css'
...
      <div className={s.header}>
        <h2 className={s.name}>
          {product.name}
        </h2>
      </div>

```

Note that if I edit my class header with   `@apply border-4 border-black border-opacity-20` it still doesn't work.

Even if I add the class using `cn` for exemple like so `className={cn(s.header, "border-opacity-0")}` nor replacing by all the style works: `className="relative bottom-0  border-opacity-0 border-t border-primary-2 rounded-b-xl h-28 flex items-center md:justify-center text-black"`

But this works `className="border-4 border-black border-opacity-20"`  while  this  does not `className="border-4 border-red border-opacity-20"`  

It feels like only the black color works with the opacity.

**Reply by adamwathan** [reply]
Hey! A Tailwind Play that reproduces your problem is all we need:

https://play.tailwindcss.com

There is no `border-red` in Tailwind by default btw, need a number like `border-red-500`.

**Reply by Florian-crg** [reply]
Ok I understand, I guess it is only that border-opacity doesn't work with custom colors I guess I would have to make a custom border-opacity. 

I will not use border-opacity, too much effort for me for no big deal. 

And your play.tailwind is not in the version I gave from the issue.

You'll see if others have the same problem I guess you're right it probably come from my setup. Thanks for your help though

**Reply by ishu202** [reply]
in my tailwind config i am overriding default colors along with the default theme colors provided by the tailwind.
this is what my color config looks like and the opacity function is same as the mentioned on the tailwind css site.
i believe it has to do something with the extended colors applying with no modifier. i will look into the library and will post soon. 

**Steps to reproduce:-** 

- create a main css file and add the colors to it inside @layers base{}
- setup tailwind config file to get those css variables like below.
- finally try applying border with opacity with class " border-primary/[.opacity] " . you will see it wouldn't work.
> hopefully this will be fixed soon.

here is the fix. the borderOpacity doesn't account for the variables passed in the colors with opacity. since the custom colors passed in the config doesn't exist in the borderOpacity it doesn't work. 
that being said if we passing an extra property `borderOpacity: ({theme}) => theme('opacity')` that works in this case. now it makes me think what else it isn't merging right in the default config object. hope this helps anyone who comes to this page.

i looked up the library it self it and all the utilities functions a lot of parsing is being done with the Regex. i believe if we use some parser to generate those lexar files like antler with node runtime. that would be less code and could be easily maintained. just some opinions i had the way its written.

hope this was helpful ✌🏼 .

now below code is generating a border around it with the same opacity of the bg-primary/75 
 link to play ground https://play.tailwindcss.com/eZuSQKsf9E?file=css

slightly hacky but should work.

```
extend: {
            colors: {
                primary: withOpacityValue('--color-primary'),
                secondary: withOpacityValue('--color-secondary'),
                success: withOpacityValue('--color-success'),
                danger: withOpacityValue('--color-danger'),
                warning: withOpacityValue('--color-warning'),
                info: withOpacityValue('--color-info'),
                light: withOpacityValue('--color-light'),
                dark: withOpacityValue('--color-dark')
            }
        }
```

New Config:-
```
extend: {
            colors: {
                primary: withOpacityValue('--color-primary'),
                secondary: withOpacityValue('--color-secondary'),
                success: withOpacityValue('--color-success'),
                danger: withOpacityValue('--color-danger'),
                warning: withOpacityValue('--color-warning'),
                info: withOpacityValue('--color-info'),
                light: withOpacityValue('--color-light'),
                dark: withOpacityValue('--color-dark')
            },
            borderOpacity: ({theme}) => theme('opacity')
        }
```

**Reply by Johnrobmiller** [reply]
Wait, so border-opacity will not work out of the box for custom colors? ):
This is almost certainly a bug, right?
So this should get moved back into "issues", or another issue should be opened up that addresses this, right?

**Reply by ishu202** [reply]
I tried explaining as much as I could. I haven't ran into any other issue so far but yes this should be fixed.

**Comment by aramrw** [comment]
> Wait, so border-opacity will not work out of the box for custom colors? ): This is almost certainly a bug, right? So this should get moved back into "issues", or another issue should be opened up that addresses this, right?

So I'm guessing this will never be added?

**Reply by brennanangel** [reply]
The correct approach is to use RGB values, see https://tailwindcss.com/docs/customizing-colors#using-css-variables
